File: 1237-find-positive-integer-solution-for-a-given-equation/1237-find-positive-integer-solution-for-a-given-equation.cpp
/*
 * // This is the custom function interface.
 * // You should not implement it, or speculate about its implementation
 * class CustomFunction {
 * public:
 *     // Returns f(x, y) for any given positive integers x and y.
 *     // Note that f(x, y) is increasing with respect to both x and y.
 *     // i.e. f(x, y) < f(x + 1, y), f(x, y) < f(x, y + 1)
 *     int f(int x, int y);
 * };
 */

class Solution {
public:
    vector<vector<int>> findSolution(CustomFunction& customfunction, int z) {
        vector<vector<int>> ans;
        for(int x = 1; x <= 1000; ++x) {
            int l = 1, r = 1000;
            while(l < r) {
                int m = (r-l)/2 + l;
                if(customfunction.f(x, m) >= z) r = m;
                else l = m+1;
            }
            if(customfunction.f(x, l) == z) ans.push_back({x, l});
        }
        return ans;
    }
};

File: 1237-find-positive-integer-solution-for-a-given-equation/1237-find-positive-integer-solution-for-a-given-equation.py
"""
   This is the custom function interface.
   You should not implement it, or speculate about its implementation
   class CustomFunction:
       # Returns f(x, y) for any given positive integers x and y.
       # Note that f(x, y) is increasing with respect to both x and y.
       # i.e. f(x, y) < f(x + 1, y), f(x, y) < f(x, y + 1)
       def f(self, x, y):
"""

class Solution(object):
    def findSolution(self, customfunction, z):
        """
        :type num: int
        :type z: int
        :rtype: List[List[int]]
        """
        ans = []
        for x in range(1, z + 1):
            # Use a simple binary search implementation without bisect.
            left, right = 1, z
            while left <= right:
                y = (left + right) // 2
                value = customfunction.f(x, y)
                if value == z:
                    ans.append([x, y])
                    break
                elif value < z:
                    left = y + 1
                else:
                    right = y - 1
        return ans
        